seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,212 @@
+# This module provides python bindings for the AnVIL Terra API.
+
+import json
+import logging
+
+from urllib.parse import urljoin
+
+from google.auth.transport.requests import AuthorizedSession
+from google.oauth2 import service_account
+
+from settings import SEQR_VERSION, TERRA_API_ROOT_URL, GOOGLE_SERVICE_ACCOUNT_INFO
+
+SEQR_USER_AGENT = "seqr/" + SEQR_VERSION
+
+scopes = ['https://www.googleapis.com/auth/userinfo.profile',
+          'https://www.googleapis.com/auth/userinfo.email',
+          'https://www.googleapis.com/auth/cloud-billing',
+          'openid']
+
+logger = logging.getLogger(__name__)
+
+
+class TerraAPIException(Exception):
+    """For exceptions happen in Terra API calls."""
+    pass
+
+
+def _seqr_agent_header(headers=None):
+    """
+    Generate seqr/version as the User-Agent header.
+
+    Args:
+        headers (dict): Include additional headers as key-value pairs
+    Returns:
+        request headers for Terra API message.
+    """
+    seqr_headers = {"User-Agent": SEQR_USER_AGENT}
+    if headers is not None:
+        seqr_headers.update(headers)
+    return seqr_headers
+
+
+class AnvilSession(AuthorizedSession):
+    def __init__(self, credentials=None, service_account_info=None, scopes=None):
+        """
+        Create an AnVIL session for a user account if credentials are provided, otherwise create one for the service account
+
+        :param credentials: User credentials
+        :param service_account_info: service account secrects
+        :param scopes: scopes of the access privilege of the session
+        """
+        if credentials is None:
+            credentials = service_account.Credentials.from_service_account_info(service_account_info, scopes = scopes)
+        super(AnvilSession, self).__init__(credentials)
+
+    def _get_call_args(self, methcall, headers=None, root_url=None):
+        if not headers:
+            headers = _seqr_agent_header()
+        if root_url is None:
+            root_url = TERRA_API_ROOT_URL
+        url = urljoin(root_url, methcall)
+        return url, headers
+
+    def get(self, methcall, headers=None, root_url=None, **kwargs):
+        """
+        Call Terra API with HTTP GET method with an authentication header.
+
+        Args:
+            methcall: A string of API path (start right after the domain name without leading slash (/)
+            headers (dict): Include additional headers as key-value pairs
+            root_url: the url up to the domain name ending with a slash (/)
+            kwargs: other parameters for the HTTP call, e.g. queries.
+        Return:
+            HTTP response
+        """
+        url, headers = self._get_call_args(self, methcall, headers = None, root_url = None)
```

Comment:
Codacy found an issue: [Argument 'headers' passed by position and keyword in method call](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6294554)